### PR TITLE
[NFC] AST: Define SubstitutionMap::getInnermostReplacementTypes

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
In preparation for storing substitution maps in bound generic types, centralize the logic for retrieving the innermost generic arguments in a new method on `SubstitutionMap`.

I wasn't able to find any duplication to clean up, so right now the only client is `TypeAliasType::getInnermostGenericArgs`. The idea is for the new method to ultimately have at least two clients, including a future `BoundGenericType::getInnermostGenericArgs`.
___
For #31895

### CODE REVIEW

File: lib/AST/SubstitutionMap.cpp [hunk 1]
```diff
@@ -97,6 +97,13 @@ ArrayRef<Type> SubstitutionMap::getReplacementTypes() const {
   return getReplacementTypesBuffer();
 }
 
+ArrayRef<Type> SubstitutionMap::getInnermostReplacementTypes() const {
+  if (empty()) { };
```
Comment (LucianoPAlmeida) [inline lib/AST/SubstitutionMap.cpp:101]: Is this correct? Just curious :)

Comment (AnthonyLatsis) [reply lib/AST/SubstitutionMap.cpp:101]: Already noticed, thank you!

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -1804,12 +1804,10 @@ class TypeAliasType final
...
-  /// arguments that are directly applied to the typealias declaration in
-  /// produced by \c getDecl().
-  ///
-  /// The result can be empty, if the declaration itself is non-generic but
-  /// the parent is generic.
-  SmallVector<Type, 2> getInnermostGenericArgs() const;
+  /// arguments that are directly applied to the typealias declaration
+  /// produced by \c getDecl(). Generic arguments applied to a parent
+  /// type are not included.
+  ArrayRef<Type> getInnermostGenericArgs() const;
```
Comment (AnthonyLatsis) [inline include/swift/AST/Types.h:1810]: I am not a fan of `getInnermostGenericArgs` though. "Direct" generic arguments and innermost generic arguments are not always the same thing.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
Ping

**Comment by AnthonyLatsis** [comment]
@slavapestov ping

**Review by slavapestov — APPROVED** [review]

**Comment by AnthonyLatsis** [comment]
Thanks!

**Pull request merged**